### PR TITLE
Improves "Extract locale variable" when working with generic types

### PR DESCRIPTION
**Pull request by SlavaRa**
Before commit: coderefactor_extract_local_variable_for_generics_bug

After: ASCompletion.Tests/Test Files/generated/haxe/AfterGenerateExtractVariableGeneric.hx

### CODE REVIEW

File: External/Plugins/ASCompletion/Completion/ASGenerator.cs [hunk 1]
```diff
@@ -2793,22 +2792,10 @@ public static void GenerateExtractVariable(ScintillaControl Sci, string NewName)
                         }
                     }
 
-                    int style = Sci.StyleAt(funcBodyStart + lastPos) & stylemask;
-                    if (ASComplete.IsCommentStyle(style))
```
Comment (elsassph) [inline External/Plugins/ASCompletion/Completion/ASGenerator.cs]: Looking generally alright, but don't we want to keep some of these conditions?

Comment (SlavaRa) [inline External/Plugins/ASCompletion/Completion/ASGenerator.cs]: The only condition is that we are interested in right now checking for comments.

Comment (elsassph) [inline External/Plugins/ASCompletion/Completion/ASGenerator.cs]: Ah right I see it later.

File: External/Plugins/ASCompletion/Completion/ASGenerator.cs [hunk 2]
```diff
@@ -2793,22 +2792,10 @@ public static void GenerateExtractVariable(ScintillaControl Sci, string NewName)
...
-
-                    }
-                    else if (!ASComplete.IsTextStyle(style))
-                    {
-                        continue;
-                    }
-
-                    Sci.SetSel(funcBodyStart + lastPos, funcBodyStart + lastPos + expression.Length);
+                    var pos = funcBodyStart + lastPos;
+                    int style = Sci.StyleAt(pos) & stylemask;
```
Comment (elsassph) [inline External/Plugins/ASCompletion/Completion/ASGenerator.cs]: `Sci.BaseStyleAt` does the masking.

Comment (SlavaRa) [inline External/Plugins/ASCompletion/Completion/ASGenerator.cs]: @elsassph, code:
```
int stylemask = (1 << Sci.StyleBits) - 1;
...
Sci.StyleAt(pos) & stylemask
```
equals to
```
Sci.BaseStyleAt(pos)
```
I have to replace "Sci.StyleAt(pos) & stylemask" to "Sci.BaseStyleAt(pos)"?

Comment (elsassph) [inline External/Plugins/ASCompletion/Completion/ASGenerator.cs]: That's the point of this function: to make the code simpler.

**Comment by Neverbirth** [comment]
Really nice to see an unit test for this PR. Do you think you could make some more? Since this is a untested generator I'm sure there are a lot of cases to add. Not asking for every possible case, of course, just the main ones that you can come of, and surely you can come up with some other failing or missing cases.

File: External/Plugins/ASCompletion/Completion/ASGenerator.cs (hunk 2 above)
Comment (SlavaRa) [inline External/Plugins/ASCompletion/Completion/ASGenerator.cs]: ok, I will refactor soon

**Comment by SlavaRa** [comment]
@Neverbirth yes, I'm going to do yet.

**Comment by SlavaRa** [comment]
@elsassph @Neverbirth guys, I have to do something for this issue? Or I can add new tests and functionality in a separate issue(for example a clear indication to the type for AS3)?

**Comment by Neverbirth** [comment]
I'd prefer to see more tests in this PR, because it will show that both the new, improved, behaviour, and the old ones, still work, and maybe fix even more things before merging.

**Comment by SlavaRa** [comment]
Ok, I add in the near future

**Comment by SlavaRa** [comment]
I added the test cases to test the extract local variable. This is a situation that I face every day on the main work.

**Comment by elsassph** [comment]
Conflict.

**Comment by SlavaRa** [comment]
@elsassph fixed.

**Pull request merged**